R/pmcmc_parameters.R

```diff
@@ -132,50 +134,142 @@ pmcmc_parameters <- R6::R6Class(
         }
+
+        if (!is.null(dimnames(proposal))) {
+          ## At this point we could reorder if that's useful and these
+          ## are setequal.
+          ok <- identical(rownames(proposal), names(parameters)) &&
+            identical(colnames(proposal), names(parameters)) &&
+            identical(dimnames(proposal)[3], populations)
+          if (!ok) {
+            stop("Expected dimension names of 'proposal' to match parameters
+            and populations.")
+          }
+        }
+
+         private$proposal <- apply(proposal, 3, rmvnorm_generator)
+         private$min <- sapply(parameters, function(.x) {
+           min <- .x[["min"]]
+           if (length(min) == 1) {
+             min <- rep(min, length(populations))
+           }
+           min
+         })
+         private$max <- sapply(parameters, function(.x) {
+           max <- .x[["max"]]
+           if (length(max) == 1) {
+             max <- rep(max, length(populations))
+           }
+           max
+         })
+         rownames(private$min) <- populations
+         rownames(private$max) <- populations
+
+      } else {
+        assert_is(proposal, "matrix")
+        # proposal
+        if (!all(dim(proposal) == length(parameters))) {
+          stop(sprintf(
+            "Expected a square proposal matrix with %d rows and columns",
+            length(parameters)))
+        }
+        if (!is.null(dimnames(proposal))) {
+          ## At this point we could reorder if that's useful and these
+          ## are setequal.
+          ok <- identical(rownames(proposal), names(parameters)) &&
+            identical(colnames(proposal), names(parameters))
+          if (!ok) {
+            stop("Expected dimension names of 'proposal' to match parameters")
+          }
+        }
+
+         private$proposal <- rmvnorm_generator(proposal)
+         private$min <- vnapply(parameters, "[[", "min",
+                                USE.NAMES = FALSE)
+         private$max <- vnapply(parameters, "[[", "max",
+                                USE.NAMES = FALSE)
+      }
+
+      if (is.null(transform)) {
+        transform <- as.list
+      } else {
+        assert_is(transform, "function")
       }
 
       private$parameters <- parameters
       private$proposal_kernel <- proposal
-      private$proposal <- rmvnorm_generator(proposal)
       private$transform <- transform
 
       private$discrete <- vlapply(private$parameters, "[[", "discrete",
                                   USE.NAMES = FALSE)
-      private$min <- vnapply(private$parameters, "[[", "min",
-                             USE.NAMES = FALSE)
-      private$max <- vnapply(private$parameters, "[[", "max",
-                             USE.NAMES = FALSE)
     },
 
     ##' @description Return the initial parameter values as a named numeric
-    ##' vector
+    ##' vector or named matrix for multiple populations.
+    # FIXME - COULD CONSIDER ALWAYS RETURNING MATRIX FOR CONSISTENCY
```

Comment:
The consistency thing here is a real trick. To me this really smells like a different type of object. The biggest issue I have with the design that we're going to have to have all our methods become these big if/else blocks that are essentially determined at construction time.

If we used a different name for this sort of object (`pmcmc_parameters_nested` or whatever) and had that _contain_ `pmcmc_parameters` objects in a list it might work better? It would match the same interface as the `pmcmc_parameters` in function names but with additional dimensions as expected

Reply:
My only concern about this is it may expect too much from the user, i.e. they may reasonably ask "once I've created a special parameter object, why do I also need a special set". I see the point with much neater code internally of course and also w.r.t. returning same outputs. 

I guess clear documentation and maybe a specialised vignette might make this clearer, e.g. "use these objects *only when*" and then it's clear they are for specific (not the general) use-case. We'd still need an assertion in `pmcmc_parameters` to check no varied parameters are included